### PR TITLE
Adding arti-rest call for iOS ffi

### PR DESCRIPTION
**Pull request by ineiti**
The rest-part of the ios-library for arti-rest. The argument-passing between ios and rest is done using a JSON string that is parsed on both ends.

See also #28

### CODE REVIEW

File: src/ffi/ios.rs [hunk 1]
```diff
@@ -4,19 +4,90 @@ use core_foundation::{
...
         .finish();
     tracing::subscriber::set_global_default(subscriber).expect("to be the only logger");
 }
 
+#[no_mangle]
+pub unsafe extern "C" fn call_arti(
+    method_c: *const c_char,
+    url_c: *const c_char,
+    _headers_c: *const c_char,
+    _body_c: *const c_char,
```
Comment (ubamrein) [inline src/ffi/ios.rs:26]:
```suggestion
    _body_c: *const u8,
```
I think this should be byte not char

Comment (ineiti) [reply src/ffi/ios.rs]: It's all JSON now...

File: src/ffi/ios.rs [hunk 2]
```diff
@@ -2,41 +2,110 @@ use core_foundation::{
...
-            "Result is: {:?}",
-            s.map(|raw| String::from_utf8(raw).expect("decode response as utf8"))
-        )),
-        Err(e) => to_cf_str(format!("Error while getting result: {}", e)),
+    ReturnStruct::new(resp.status(), resp.headers(), resp.body())
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct ReturnStruct {
+    error: String,
```
Comment (ubamrein) [inline src/ffi/ios.rs:78]: I would suggest something like:

```rust
pub struct ReturnStruct {
   error: Option<Error>,
   response: Option<Response>
}

pub struct Response {
  status: u16,
  headers: HashMap<String, Vec<String>>,
 body: Vec<u8>
}

pub struct Error {
   error_string: String,
  error_context: Option<String>
}

impl ReturnStruct {

pub fn ok(status: StatusCode, headers_map: &HeaderMap<HeaderValue>, body_vec: &Vec<u8>) -> Result<Self> {
    let mut headers = HashMap::new();
        for (k, v) in headers_map {
            headers.insert(k.to_string(), vec![v.to_str()?.to_string()]);
        }

      Ok(Self {
         error: None,
          response: Some(Response {
           status,
          headers,
         body: body_vec.clone()
        })
    })

 pub fn error(error: &str) -> Self {
    Self {
     error: Some(Error {
          error: error.to_string(),
         error_context: None //or whatever
     }),
    response: None
 }
// plus to_json or whatever functions
// maybe a is_error() and is_ok() or so?
}
```

File: src/ffi/ios.rs [hunk 3]
```diff
@@ -2,41 +2,110 @@ use core_foundation::{
...
+        "POST" => Request::post(request.url),
+        "PUT" => Request::put(request.url),
+        "DELETE" => Request::delete(request.url),
+        // TODO: add other verbs of REST
+        // TODO: correctly return error here
+        _ => Request::get(request.url),
+    }
+        .header("Host", host)
+        .version(http::Version::HTTP_10)
+        .body(vec![])?;
```
Comment (tharvik) [inline src/ffi/ios.rs:54]: I found that you can use `http::method::Method::from_bytes` to automatically parse the method.
I've also added a quick example on how to better handle error: don't use `unwrap` or a `?` without `context`. `anyhow::Context` adds `context()` on both `Option` and `Result` so you can easily add some info before returning.

```suggestion
    let host = Url::parse(&request.url)
        .context("parse url")?
        .host_str()
        .context("no host in request")?;

    let mut req = Request::builder()
        .method(Method::from_bytes(request.method.as_bytes()).context("invalid method")?)
        .header("Host", host)
        .version(http::Version::HTTP_10)
        .body(vec![])
        .context("invalid request")?;
```

File: src/ffi/ios.rs [hunk 4]
```diff
@@ -2,41 +2,110 @@ use core_foundation::{
...
-pub unsafe extern "C" fn call_tls_get(domain_cc: *const c_char) -> CFStringRef {
-    let domain = cstring_to_str(&domain_cc);
+pub unsafe extern "C" fn call_arti(
+    request_json: *const c_char) -> CFStringRef {
+    setup_logger();
+
+    let ret = match _call_arti(cstring_to_str(&request_json)) {
+        Ok(ret) => ret,
+        Err(e) => ReturnStruct::err(e.to_string()),
+    };
```
Comment (tharvik) [inline src/ffi/ios.rs:34]: `unwrap_or_else` returns the `Ok` value or execute a function when it's an `Err` before returning.

```suggestion
    let ret = _call_arti(cstring_to_str(&request_json))
        .unwrap_or_else(|e| ReturnStruct::err(e.to_string()));
```

File: src/ffi/ios.rs [hunk 5]
```diff
@@ -2,41 +2,110 @@ use core_foundation::{
...
-        )),
-        Err(e) => to_cf_str(format!("Error while getting result: {}", e)),
+    ReturnStruct::new(resp.status(), resp.headers(), resp.body())
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct ReturnStruct {
+    error: String,
+    status: u16,
+    headers: HashMap<String, Vec<String>>,
```
Comment (tharvik) [inline src/ffi/ios.rs:80]: Why is it using a `Vec<String>` as values? From what I remember, it was quite a corner-case (and undefinied in the HTTP spec) to have multiple values for a single header-nam.

File: src/ffi/ios.rs [hunk 6]
```diff
@@ -2,41 +2,110 @@ use core_foundation::{
...
+    body: Vec<u8>,
+}
+
+impl ReturnStruct {
+    pub fn new(status: StatusCode, headers_map: &HeaderMap<HeaderValue>, body_vec: &Vec<u8>) -> Result<Self> {
```
Comment (tharvik) [inline src/ffi/ios.rs:85]: You can direclty create it from the `http::Request`.

NB: if you want to make it even nicer, you can remove `new` and `err` and instead implement `std::convert::TryFrom<Result<http::Request>>`

File: src/ffi/ios.rs (hunk 5 above)
Comment (ubamrein) [reply src/ffi/ios.rs:80]: It is not a corner case at all :) Just make a request, where there are cookies being set :) It is also defined in the HTTP-Spec.

>  Multiple message-header fields with the same field-name MAY be
   present in a message if and only if the entire field-value for that
   header field is defined as a comma-separated list [i.e., #(values)].

[RFC-2616](https://datatracker.ietf.org/doc/html/rfc2616#page-31])

Comment (tharvik) [reply src/ffi/ios.rs:80]: Oupsi, I inversed the key and the values :sweat_smile: thanks for catching it.

File: src/ffi/ios.rs (hunk 2 above)
Comment (ineiti) [reply src/ffi/ios.rs]: Looks good, thanks

File: src/arti.rs [hunk 7]
```diff
@@ -96,11 +96,13 @@ async fn get_tor<T: Runtime>(
...
-    debug!(?cache_path);
-
     let mut dircfg = tor_dirmgr::NetDirConfigBuilder::new();
-    dircfg.set_cache_path(cache_path);
+    if let Some(cache_path) = cache_dir {
+        debug!(?cache_path);
+        dircfg.set_cache_path(std::path::Path::new(cache_path));
+    } else {
+        dircfg.use_default_cache_path()?;
+    }
```
[the remark this answers is not in the document]
Comment (ineiti) [reply src/arti.rs:105]: Guess so, perhaps others, too. For me it's OK to merge #38, then I'll rebase on top of that.

[the remark this answers is not in the document]
Comment (tharvik) [reply src/arti.rs:105]: Done :)

File: Cargo.toml [hunk 8]
```diff
@@ -65,7 +66,8 @@ env_logger = "0.6"
 tempdir = "0.3"
 
 [profile.release]
+incremental=true
+codegen-units=256
+panic = "abort"
```
[the remark this answers is not in the document]
Comment (ineiti) [reply Cargo.toml:71]: rebase took care of that.

File: src/ffi/ios.rs (hunk 6 above)
Comment (ineiti) [reply src/ffi/ios.rs]: I tried to implement it here, but it still fails and I don't understand the compiler-error :(

Comment (ineiti) [reply src/ffi/ios.rs]: Now it works - thanks!

File: src/ffi/ios.rs [hunk 9]
```diff
@@ -1,40 +1,130 @@
...
+                    error_context: None,
+                }),
+                response: None,
+            }),
+        }
+    }
+}
+
+impl ReturnStruct {
+    pub fn err(error: String) -> ReturnStruct {
```
Comment (tharvik) [inline src/ffi/ios.rs:116]: You can remove it now, and use `try_from` directly

File: src/ffi/ios.rs [hunk 10]
```diff
@@ -1,40 +1,130 @@
...
+        ReturnStruct {
+            error: Some(JSONError {
+                error_string: error.to_string(),
+                error_context: None,
+            }),
+            response: None,
+        }
+    }
+
+    pub fn to_json(&self) -> String {
```
Comment (tharvik) [inline src/ffi/ios.rs:126]: You can also inline this single call, thus separating concerns of serialization from the ReturnStruct's implementation.

And yeah, I'm all for reducing the number of exposed functions, the less there is, the safer it is IMO :)

File: Cargo.toml [hunk 11]
```diff
@@ -26,11 +26,13 @@ httparse = "1"
 log = "0.4"
 log-panics = "2"
 serde = { version = "1", features = ["derive"] }
+serde_json = "1.0.64"
 tokio-rustls = "0.22"
 tokio = { version = "1", features = ["full"] }
 tracing = "0.1"
 tracing-fmt = "0.1"
 tracing-log = "0.1"
+url = "2.2.2"
```
Comment (tharvik) [inline Cargo.toml:35]: You can move it to the iOS specific deps

**Review by tharvik — APPROVED** [review]
Some cosmetic changes, otherwise, LGTM

**Comment by ineiti** [comment]
Applied latest comments, but put it back into `In Progress`, as it needs some comments in the code...

**Comment by ineiti** [comment]
@tharvik as I won't touch this before next Monday, let's merge it anyway, and I'll work on the comments next week...

If you're OK with the changes in here, I let you click the `Rebase and merge` button.

**Comment by tharvik** [comment]
> @tharvik as I won't touch this before next Monday, let's merge it anyway, and I'll work on the comments next week...
> If you're OK with the changes in here, I let you click the `Rebase and merge` button.

Okay, I'll add my commits directly and merge it then.

**Pull request merged**